doc/api/metrics.md

```diff
@@ -0,0 +1,1040 @@
+metrics that can be collected and exported to monitoring systems. All metrics
+publish their data through the `node:diagnostics_channel` module, allowing for
+flexible consumption patterns.
+
+### Example
+
+```mjs
+import { counter, timer, statsdStream } from 'node:metrics';
+import { createWriteStream } from 'node:fs';
+
+// Create a counter metric
+const apiCalls = counter('api.calls', { service: 'web' });
+
+// Create a timer factory
+const requestTimer = timer('api.request.duration', { service: 'web' });
+
+// Export metrics to StatsD format
+const statsd = statsdStream();
+statsd.pipe(createWriteStream('metrics.log'));
+
+// Use metrics in your application
+function handleRequest(req, res) {
+  const timer = requestTimer.create({ endpoint: req.url });
+
+  apiCalls.increment();
+
+  // Process request...
+
+  timer.stop();
+}
+```
+
+```cjs
+const { counter, timer, statsdStream } = require('node:metrics');
+const { createWriteStream } = require('node:fs');
+
+// Create a counter metric
+const apiCalls = counter('api.calls', { service: 'web' });
+
+// Create a timer factory
+const requestTimer = timer('api.request.duration', { service: 'web' });
+
+// Export metrics to StatsD format
+const statsd = statsdStream();
+statsd.pipe(createWriteStream('metrics.log'));
+
+// Use metrics in your application
+function handleRequest(req, res) {
+  const timer = requestTimer.create({ endpoint: req.url });
+
+  apiCalls.increment();
+
+  // Process request...
+
+  timer.stop();
+}
+```
+
+## Metric Types
+
+### `metrics.counter(name[, meta])`
+
+<!-- YAML
+added: REPLACEME
+-->
+
+* `name` {string} The name of the counter metric.
+* `meta` {Object} Optional metadata to attach to all reports.
+* Returns: {metrics.Counter}
+
+Creates a counter metric that tracks cumulative values.
+
+```mjs
+import { counter } from 'node:metrics';
+
+const errorCount = counter('errors.total', { component: 'database' });
+
+errorCount.increment();     // Increment by 1
+errorCount.increment(5);    // Increment by 5
+errorCount.decrement(2);    // Decrement by 2
+```
+
+
+### `metrics.gauge(name[, meta])`
+
+<!-- YAML
+added: REPLACEME
+-->
+
+* `name` {string} The name of the gauge metric.
+* `meta` {Object} Optional metadata to attach to all reports.
+* Returns: {metrics.Gauge}
+
+Creates a gauge metric that represents a single value at a point in time.
+
+```mjs
+import { gauge } from 'node:metrics';
+import { memoryUsage } from 'node:process';
+
+const memory = gauge('memory.usage.bytes');
+
+memory.reset(memoryUsage().heapUsed);
+memory.applyDelta(1024);  // Add 1024 to current value
+```
+
+### `metrics.meter(name, interval[, meta])`
+
+<!-- YAML
+added: REPLACEME
+-->
+
+* `name` {string} The name of the meter metric.
+* `interval` {number} The time window in milliseconds for rate calculation.
+* `meta` {Object} Optional metadata to attach to all reports.
+* Returns: {metrics.Meter}
+
+Creates a meter metric that measures the rate of events over time.
```

Comment:
What does the "rate" here refer to? I think this meter type needs more explanation as it is not a common type in existing metric backends.

Reply:
Ah, I misunderstood, it's actually not part of vanilla statsd apparently, but rather is included as an addition in basically every implementation of statsd. In dogstatsd it's called rate, in Sysdig it's called meter, most other APM vendors have some variation of it.

In the case of Prometheus, it doesn't make sense because the pull model keeps the data in-app, and having that unaggregated would be prohibitively expensive.

Reply:
This sounds like the API design should still be a counter. In the exportation, they can decide if they would like to aggregate in a time interval, or until being pull-ed.

Reply:
They are generally marked as separate types right from the initial API as the backend usually needs intent communicated to know which aggregation method to apply. Some just apply all the aggregations though and make every representation available, though that can get very expensive for storage.

Reply:
To clarify my comment at https://github.com/nodejs/node/pull/58874#discussion_r2174902336. I think it is fine to add more semantic attributes to a metric. If a backend needs more semantics, they can even be added as meta.

> Applying an interval and aggregation in the metric collection adds additional hurdles when there are exporters configured with different exportation intervals. They have to handle the aggregation with different size of time windows.

What's really concerning is doing aggregation in the metric collection. This practically prohibits some backend exporters to be implemented efficiently.